pkg/alertmanager/state_persister.go

```diff
@@ -0,0 +1,100 @@
+package alertmanager
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/alertmanager/cluster/clusterpb"
+
+	"github.com/cortexproject/cortex/pkg/alertmanager/alertspb"
+	"github.com/cortexproject/cortex/pkg/alertmanager/alertstore"
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+const (
+	defaultPersistInterval = 15 * time.Minute
+	defaultPersistTimeout  = 30 * time.Second
+)
+
+type PersistableState interface {
+	State
+	GetFullState() (*clusterpb.FullState, error)
+}
+
+// statePersister periodically writes the alertmanager state to persistent storage.
+type statePersister struct {
+	services.Service
+
+	state  PersistableState
+	store  alertstore.AlertStore
+	userID string
+	logger log.Logger
+
+	interval time.Duration
+	timeout  time.Duration
+}
+
+// newStatePersister creates a new state persister.
+func newStatePersister(userID string, state PersistableState, store alertstore.AlertStore, l log.Logger) *statePersister {
```

Comment:
I think we need some metrics to track the health of all state persisters. Eg. state persistence attempts + failures counter. Without metrics we can't alert on failures in the persister.

Reply:
Agreed. I will add these in a separate PR to keep the size of this PR small, and I would also like to add some other metrics relating to the state (e.g. number of times we had read the state from a replica. number of times reading from storage, and so forth).